opendevin/controller/agent_controller.py

```diff
@@ -196,25 +217,46 @@ async def _step(self):
             await asyncio.sleep(1)
             return
 
-        logger.info(f'STEP {self.state.iteration}', extra={'msg_type': 'STEP'})
-        if self.state.iteration >= self.max_iterations:
-            await self.report_error('Agent reached maximum number of iterations')
-            await self.set_agent_state_to(AgentState.ERROR)
-            return
-
         if self.delegate is not None:
-            delegate_done = await self.delegate._step()
+            await self.delegate._step()
+            delegate_state = self.delegate.get_agent_state()
+            if delegate_state == AgentState.ERROR:
```

Comment:
I was trying this out, and managed to hit this error:
```11:47:36 - opendevin:ERROR: agent_controller.py:125 - Traceback (most recent call last):
  File "/home/rbren/git/opendevin/opendevin/controller/agent_controller.py", line 118, in _start_step_loop
    await self._step()
  File "/home/rbren/git/opendevin/opendevin/controller/agent_controller.py", line 220, in _step
    delegate_state = self.delegate.get_agent_state()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get_agent_state'

```

My guess is that `_step()` sends a Finish action, which makes a callback to the controller, which unsets `self.delegate`

Reply:
Reading the code some more, I'm actually very confused as to how this happened. I do have several delegate layers going

Reply:
Hey do you have an example command that his this error?

Reply:
Unfortunately no--I haven't been able to repro it. Maybe we can ignore for now...

Reply:
I just hit it again ☹️ let me see if I can create a consistent repro

Reply:
Thanks for spotting this bug! Turns out the curse was `_step()` function called by both `_start_step_loop` and parent agent controller. I fixed by NOT starting step loop for delagates.